src/oom/oomd-util.c

```diff
@@ -0,0 +1,379 @@
+                        target = key;
+                }
+
+        if (curr_mem_pressure.avg10 > threshold) {
+                if (*ret_last_hit_memory_threshold == 0)
+                        *ret_last_hit_memory_threshold = now(CLOCK_MONOTONIC);
+
+                diff = now(CLOCK_MONOTONIC) - *ret_last_hit_memory_threshold;
+                if (diff >= duration)
+                        return target;
+        } else
+                *ret_last_hit_memory_threshold = 0;
+
+        return NULL;
+}
+
+bool oomd_memory_reclaim(Hashmap *h) {
+        uint64_t pgscan = 0, pgscan_of = 0, last_pgscan = 0, last_pgscan_of = 0;
+        OomdCGroupContext *ctx;
+        Iterator i;
+
+        if (!h)
+                return false;
+
+        HASHMAP_FOREACH(ctx, h, i) {
+                if (pgscan > UINT64_MAX - ctx->pgscan)
+                        pgscan_of++;
+                pgscan += ctx->pgscan;
+
+                if (last_pgscan > UINT64_MAX - ctx->last_pgscan)
+                        last_pgscan_of++;
+                last_pgscan += ctx->last_pgscan;
+        }
+
+        if (last_pgscan_of == pgscan_of)
+                return pgscan > last_pgscan;
+
+        return pgscan_of > last_pgscan_of;
+}
+
+bool oomd_swap_free_below(const OomdSystemContext *ctx, uint64_t threshold_percent) {
+        uint64_t swap_threshold;
+
+        assert(ctx);
+        assert(threshold_percent <= ((uint64_t) 100));
+
+        swap_threshold = ctx->swap_total * threshold_percent / ((uint64_t) 100);
+        return (ctx->swap_total - ctx->swap_used) < swap_threshold;
+}
+
+/* Get an array of OomdCGroupContexts from h, sorted from largest to smallest values according to compare_func.
+ * If prefix is not NULL, only include contexts whose cgroup paths start with prefix. Otherwise all paths are sorted. */
+int oomd_sort_cgroup_contexts(Hashmap *h, oomd_compare_t compare_func, const char *prefix, OomdCGroupContext ***ret) {
+        _cleanup_free_ OomdCGroupContext **sorted = NULL;
+        OomdCGroupContext *item;
+        size_t k = 0;
+        Iterator i;
+
+        assert(h);
+        assert(compare_func);
+        assert(ret);
+
+        sorted = new0(OomdCGroupContext*, hashmap_size(h));
+        if (!sorted)
+                return -ENOMEM;
+
+        HASHMAP_FOREACH(item, h, i) {
+                if (prefix && !startswith(item->path, prefix))
+                        continue;
+
+                sorted[k++] = item;
+        }
+
+        typesafe_qsort(sorted, k, compare_func);
+
+        *ret = TAKE_PTR(sorted);
+        return 0;
+}
+
+int oomd_kill_by_pgscan(Hashmap *h, const char *prefix, bool dry_run) {
+        _cleanup_free_ OomdCGroupContext **sorted = NULL;
+        int r;
+
+        assert(h);
+
+        r = oomd_sort_cgroup_contexts(h, compare_pgscan, prefix, &sorted);
+        if (r < 0)
+                return r;
+
+        for (size_t i = 0; i < hashmap_size(h); i++) {
+                if (sorted[i]->pgscan == 0)
+                        break;
```

Comment:
hmm, so in `oomd_sort_cgroup_contexts()` you iterate through all hashmap entries and filter some of them out, i.e. the returned array possibly has fewer entries than the hashtable has entries. Yet, you assume they have the same size here, and if they don't we get a NULL pointer deref here?

Reply:
The array was initialized to `hashmap_size(h)` and zero filled so this is fine. But since I updated `oomd_sort_cgroup_contexts` to return the number of sorted elements I figure this will be less ambiguous to use that than `hashmap_size(h)`.